### PR TITLE
fix(moe): support calibrated NVFP4 global scales

### LINKED ISSUE #3548: [Bug] Unified MoE API (#3093) only supports activation global_scale=1.0 — calibrated checkpoint scales silently inflate output (~2400x)
**Affected:** unified MoE API, PR #3093 (`moe_api` branch), `flashinfer/fused_moe/prepare.py` + `MoEActivationPack`. SM100, NVFP4.

### Summary
The MVP marshalling is wired **only for `global_scale = 1.0`**: `MoEActivationPack` has no field for the activation global scale, and `prepare_cute_dsl_nvfp4_weights` hardcodes `gs = torch.tensor([1.0])`, `fc2_input_scale = 1.0`, `alpha = ones`. As a result a caller cannot supply the **checkpoint-calibrated** scales real deployments use (vLLM/SGLang fold `w2_weight_scale * w2_input_scale` into `g2_alphas` / `output*_scale_scalar`).

It's also a **silent footgun**: if you quantize the activation with a real global scale — `a1_gs = (448*6)/max|x| ≈ 2.6e3`, i.e. what a calibrated checkpoint carries — the scale never reaches the kernel (no Pack field), and the output **inflates by roughly the activation global scale (~2400× for typical magnitudes)** with no error raised. Setting `a1_gs = 1.0` makes it correct.

### Why it matters
Not a correctness bug *at* `gs = 1.0` (block scales absorb a well-conditioned range), but it (a) blocks running a real ModelOpt/NVFP4 checkpoint through the unified API, and (b) is the exact surface of the recurring low-precision-MoE accuracy issues around small/calibrated intermediate scales.

### Repro
Drive `MoELayer` (NVFP4) from bf16 weights and build the activation pack with `a1_gs = (448*6)/x.abs().max()` instead of `1.0`. Output is ~2400× the reference (both backends still agree *with each other* — consistently wrong). With `a1_gs = 1.0` it matches the reference (≈ fp4 envelope).

### Suggested fix (matches the design doc's role-named QuantSpec, §4.3)
Add `hidden_states_scale_global` (and optionally per-expert) to `MoEActivationPack` and thread it into each runner's fc1 dequant; let `prepare_weights` accept calibrated weight + intermediate scales instead of hardcoding `1.0`/`ones`. Until then, document that the MVP is `global_scale = 1.0`-only so integrators don't pass calibrated scales and get silently-scaled output.

(Found by a unified-API MoE fuzzer; file:line refs at PR #3093 head `89466b1a`.)

### PR DESCRIPTION
**Pull request by XFDG**
## Description

The unified SM100 NVFP4 MoE path quantized activations and weights with a hard-coded global scale of 1.0. Supplying checkpoint-calibrated activation data therefore lost its global scale before GEMM1 and could inflate output by roughly that scale.

This change:

- adds keyword-only `hidden_states_scale_global` metadata to `MoEActivationPack`;
- lets TRTLLM, Cake, and CuteDSL activation/weight preparation accept calibrated NVFP4 scales;
- supports scalar or per-expert GEMM1/GEMM2 weight scales plus a scalar intermediate scale;
- folds the scales using the canonical formulas:
  - FC1 output: `c_gs / (w1_gs * a1_gs)`
  - FC1 gate: `1 / (w1_gs * a1_gs)`
  - FC2 output: `1 / (c_gs * w2_gs)`
- preserves the existing flattened quantization path and tensors when scales are omitted;
- rejects calibrated-scale arguments for non-NVFP4xNVFP4 pairs instead of silently ignoring them.

This ports the intent of closed PR #3592 onto the current typed-activation, multi-format unified MoE API.

## Related Issues

Fixes #3548.

### Pre-commit Checks

- [ ] Full `pre-commit run --all-files` was not run.
- [x] `git diff --check` passes.

## Tests

- [x] Regression coverage was added.
- [x] Focused tests pass on NVIDIA B200 (SM100), CUDA 13.1, PyTorch 2.11 nightly.

```text
pytest -q tests/moe/test_unified_moe_global_scale.py -k cute_dsl -s
1 passed, 1 deselected

pytest -q tests/moe/test_unified_moe_global_scale.py -k trtllm_fp4_routed -s
1 passed, 1 deselected
```

The test uses a non-unit calibrated activation scale, per-expert weight scales, and a non-unit intermediate scale, then compares each backend against a BF16 reference. TRTLLM cubins were downloaded on the network-enabled host and verified against the module's official SHA-256 manifest before offline B200 execution.

## Experimental Track

- [ ] This PR does not modify an experimental API or backend.

## Reviewer Notes

The default `None` scale values retain the prior global-scale-1.0 behavior. The activation scalar is folded at pack time because it is per-call metadata, while weight and intermediate factors remain in the long-lived prepared weight view.

## Summary by CodeRabbit

* **New Features**
  * Added support for calibrated global scales for NVFP4 weights and activations across supported MoE backends.
  * Calibrated scales now influence quantization and output scaling.
  * Added activation-scale support to activation preparation and packing APIs.

* **Bug Fixes**
  * Added validation for scale values, including positivity, finiteness, device placement, and supported quantization combinations.
  * Preserved previous behavior when calibrated scales are not supplied.

* **Tests**
  * Added GPU coverage comparing calibrated NVFP4 MoE results with a float32 reference implementation.

### CODE REVIEW

File: flashinfer/fused_moe/runners.py [hunk 1]
```diff
@@ -660,6 +660,23 @@ def get_cache_key_extras(self, inputs: List[torch.Tensor]) -> tuple:
...
+    """Fold the activation global scale into TRTLLM/Cake GEMM1 scalars."""
+    output1 = view.get("output1_scale_scalar")
+    output1_gate = view.get("output1_scale_gate_scalar")
+    if act.hidden_states_scale_global is None:
+        return output1, output1_gate
+    inv_scale = 1.0 / act.hidden_states_scale_global
+    if output1 is not None:
+        output1 = (output1 * inv_scale).contiguous()
+    if output1_gate is not None:
+        output1_gate = (output1_gate * inv_scale).contiguous()
```
[the remark this answers is not in the document]
Comment (XFDG) [reply flashinfer/fused_moe/runners.py]: Fixed in 8c20f037. The shared helper now revalidates same-device float32, singleton, finite, positive scale metadata and reshapes it to [1] before folding. Added mutation and singleton-shape regression coverage.

File: flashinfer/fused_moe/runners.py [hunk 2]
```diff
@@ -3991,6 +4012,18 @@ def pack_inputs(
...
+            if self.config.quant.pair != (
+                QuantFormat.NVFP4,
+                QuantFormat.NVFP4,
+            ):
+                raise ValueError(
+                    "hidden_states_scale_global is supported only for NVFP4×NVFP4."
+                )
+            w1_alpha = (
+                w1_alpha / act.hidden_states_scale_global.to(w1_alpha.device)
+            ).contiguous()
```
[the remark this answers is not in the document]
Comment (XFDG) [reply flashinfer/fused_moe/runners.py]: Fixed in 8c20f037. The per-token NVFP4 branch now passes the folded w1_alpha. The new non-unit global-scale per-token case passes on B200/SM100.

File: tests/moe/test_unified_moe_global_scale.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+@pytest.mark.parametrize(
+    "backend_config,backend_key",
+    [
+        (CuteDslConfig(), "cute_dsl"),
+        (TrtllmFp4Config(), "trtllm_fp4_routed"),
+    ],
+)
+def test_nvfp4_calibrated_global_scales_match_reference(
+    backend_config, backend_key
+):
```
[the remark this answers is not in the document]
Comment (XFDG) [reply tests/moe/test_unified_moe_global_scale.py]: Ruff formatting was fixed in 118c60fc and the follow-up remains clean under ruff format --check and ruff check.

File: tests/moe/test_unified_moe_global_scale.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            fc1 = torch.mv(w1[expert].float(), x[token].float())
+            activated = torch.nn.functional.silu(
+                fc1[intermediate_size:]
+            ) * fc1[:intermediate_size]
+            reference[token] += topk_weights[token, slot] * torch.mv(
+                w2[expert].float(), activated
+            )
+
+    assert torch.isfinite(output).all()
+    torch.testing.assert_close(output.float(), reference, rtol=0.5, atol=0.12)
```
[the remark this answers is not in the document]
Comment (XFDG) [reply tests/moe/test_unified_moe_global_scale.py]: Added in 8c20f037 using the supported Cake geometry H=2048, I=1536, E=60, top-k=4 with PDL enabled and a non-unit activation scale. The BF16-reference comparison passes on B200/SM100.

**Comment by XFDG** [comment]
Addressed the latest outside-diff review finding in commit df02076d.

The mutable hidden_states_scale_global metadata is now validated by one shared helper: same device as hidden_states_q, FP32, exactly one element reshaped to [1], finite, and positive. Both TRTLLM/Cake folding and CuteDslRunner.pack_inputs now use that helper, so the CuTe DSL path can no longer bypass revalidation.

Validation:
- B200/SM100 focused mutation and singleton-shape regressions: 5 passed
- Ruff check and format check: passed
- The full B200 file additionally reached 6 passed; its 3 backend execution cases were blocked before kernel execution by the GPU node's timeout fetching NVIDIA's batched_gemm/checksums.txt artifact.
